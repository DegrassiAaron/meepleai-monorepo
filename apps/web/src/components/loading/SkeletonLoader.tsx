/**
 * SkeletonLoader Component
 *
 * Loading placeholder with variant-specific dimensions matching real content.
 * Uses CSS-based animation (Tailwind animate-pulse) for performance.
 * Respects user's reduced motion preferences.
 *
 * @example
 * ```tsx
 * // Games grid
 * <SkeletonLoader variant="games" count={6} />
 *
 * // Agents list
 * <SkeletonLoader variant="agents" count={3} />
 *
 * // Chat messages
 * <SkeletonLoader variant="message" count={2} />
 *
 * // Chat history sidebar
 * <SkeletonLoader variant="chatHistory" count={10} />
 * ```
 */

import { useReducedMotion } from '@/lib/animations';

export interface SkeletonLoaderProps {
  /**
   * Visual variant matching the content type
   */
  variant: 'games' | 'agents' | 'message' | 'chatHistory';

  /**
   * Number of skeleton placeholders to render
   * @default 1
   */
  count?: number;

  /**
   * Additional CSS classes
   */
  className?: string;

  /**
   * Enable animation (respects prefers-reduced-motion)
   * @default true
   */
  animate?: boolean;

  /**
   * Accessible label for screen readers
   * @default 'Loading...'
   */
  ariaLabel?: string;
}

/**
 * Variant-specific styles matching real content dimensions
 */
const VARIANT_STYLES = {
  games: 'h-64 rounded-lg', // Card layout: image + title + description
  agents: 'h-20 rounded-md', // List item: icon + name + description
  message: 'h-16 rounded-xl', // Chat message: avatar + text
  chatHistory: 'h-12 rounded-md', // Compact list: title + timestamp
} as const;

/**
 * SkeletonLoader component for loading states
 */
export function SkeletonLoader({
  variant,
  count = 1,
  className = '',
  animate = true,
  ariaLabel = 'Loading...',
}: SkeletonLoaderProps) {
  const shouldReduceMotion = useReducedMotion();
  const shouldAnimate = animate && !shouldReduceMotion;

  // Generate array for rendering multiple skeletons
  const skeletons = Array.from({ length: count }, (_, index) => index);

  return (
    <div className="space-y-4">
      {skeletons.map((index) => (
        <div
          key={index}
          role="status"
          aria-live="polite"
          aria-label={ariaLabel}
          className={`
            ${VARIANT_STYLES[variant]}
            ${shouldAnimate ? 'animate-pulse' : ''}
            bg-slate-200 dark:bg-slate-700
            ${className}
          `}
        >
          {/* Screen reader text */}
          <span className="sr-only">{ariaLabel}</span>

          {/* Variant-specific content structure */}
          {variant === 'games' && (
            <div className="h-full flex flex-col p-4 space-y-3">
              {/* Image placeholder */}
              <div className="flex-1 bg-slate-300 dark:bg-slate-600 rounded" />
              {/* Title placeholder */}
              <div className="h-4 bg-slate-300 dark:bg-slate-600 rounded w-3/4" />
              {/* Description placeholder */}
              <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full" />
              <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-5/6" />
            </div>
          )}

          {variant === 'agents' && (
            <div className="h-full flex items-center p-3 space-x-3">
              {/* Icon placeholder */}
              <div className="w-12 h-12 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0" />
              {/* Text content */}
              <div className="flex-1 space-y-2">
                <div className="h-4 bg-slate-300 dark:bg-slate-600 rounded w-1/2" />
                <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4" />
              </div>
            </div>
          )}

          {variant === 'message' && (
            <div className="h-full flex items-start p-3 space-x-3">
              {/* Avatar placeholder */}
              <div className="w-10 h-10 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0" />
              {/* Message content */}
              <div className="flex-1 space-y-2 pt-1">
                <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full" />
                <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-4/5" />
              </div>
            </div>
          )}

          {variant === 'chatHistory' && (
            <div className="h-full flex items-center p-3 space-x-2">
              {/* Icon placeholder */}
              <div className="w-6 h-6 bg-slate-300 dark:bg-slate-600 rounded flex-shrink-0" />
              {/* Title and timestamp */}
              <div className="flex-1 space-y-1">
                <div className="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4" />
                <div className="h-2 bg-slate-300 dark:bg-slate-600 rounded w-1/3" />
              </div>
            </div>
          )}
        </div>
      ))}
    </div>
  );
}
