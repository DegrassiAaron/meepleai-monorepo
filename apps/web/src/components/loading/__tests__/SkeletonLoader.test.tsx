/**
 * Tests for SkeletonLoader component
 * Loading placeholders with variant-specific styles
 */

import { render, screen } from '@testing-library/react';
import { SkeletonLoader } from '../SkeletonLoader';

// Mock useReducedMotion hook
const mockUseReducedMotion = jest.fn();
jest.mock('@/lib/animations', () => ({
  ...jest.requireActual('@/lib/animations'),
  useReducedMotion: () => mockUseReducedMotion(),
}));

describe('SkeletonLoader', () => {
  beforeEach(() => {
    mockUseReducedMotion.mockReturnValue(false);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Count prop', () => {
    it('should render correct number of skeletons based on count prop', () => {
      const { container } = render(<SkeletonLoader variant="games" count={3} />);
      const skeletons = container.querySelectorAll('[role="status"]');
      expect(skeletons).toHaveLength(3);
    });

    it('should render single skeleton when count is 1', () => {
      const { container } = render(<SkeletonLoader variant="games" count={1} />);
      const skeletons = container.querySelectorAll('[role="status"]');
      expect(skeletons).toHaveLength(1);
    });

    it('should default to count of 1 when not provided', () => {
      const { container } = render(<SkeletonLoader variant="games" />);
      const skeletons = container.querySelectorAll('[role="status"]');
      expect(skeletons).toHaveLength(1);
    });
  });

  describe('Variant-specific styles', () => {
    it('should apply games variant styles (card layout)', () => {
      const { container } = render(<SkeletonLoader variant="games" />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('h-64'); // Card height
      expect(skeleton).toHaveClass('rounded-lg');
    });

    it('should apply agents variant styles (list item layout)', () => {
      const { container } = render(<SkeletonLoader variant="agents" />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('h-20'); // List item height
      expect(skeleton).toHaveClass('rounded-md');
    });

    it('should apply message variant styles (chat message layout)', () => {
      const { container } = render(<SkeletonLoader variant="message" />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('h-16'); // Message height
      expect(skeleton).toHaveClass('rounded-xl');
    });

    it('should apply chatHistory variant styles (sidebar list)', () => {
      const { container } = render(<SkeletonLoader variant="chatHistory" />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('h-12'); // Compact list item
      expect(skeleton).toHaveClass('rounded-md');
    });
  });

  describe('Animation', () => {
    it('should have animate-pulse class when animate prop is true', () => {
      const { container } = render(<SkeletonLoader variant="games" animate={true} />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('animate-pulse');
    });

    it('should not have animate-pulse class when animate prop is false', () => {
      const { container } = render(<SkeletonLoader variant="games" animate={false} />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).not.toHaveClass('animate-pulse');
    });

    it('should respect prefers-reduced-motion and disable animation', () => {
      mockUseReducedMotion.mockReturnValue(true);
      const { container } = render(<SkeletonLoader variant="games" animate={true} />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).not.toHaveClass('animate-pulse');
    });

    it('should default to animated state', () => {
      const { container } = render(<SkeletonLoader variant="games" />);
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('animate-pulse');
    });
  });

  describe('Custom className', () => {
    it('should apply custom className', () => {
      const { container } = render(
        <SkeletonLoader variant="games" className="custom-skeleton" />
      );
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('custom-skeleton');
    });

    it('should preserve base classes with custom className', () => {
      const { container } = render(
        <SkeletonLoader variant="games" className="text-red-500" />
      );
      const skeleton = container.querySelector('[role="status"]');
      expect(skeleton).toHaveClass('text-red-500');
      expect(skeleton).toHaveClass('bg-slate-200'); // Base class
    });
  });

  describe('Accessibility', () => {
    it('should have correct ARIA attributes', () => {
      render(<SkeletonLoader variant="games" ariaLabel="Loading games" />);
      const skeleton = screen.getByRole('status');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveAttribute('aria-live', 'polite');
      expect(skeleton).toHaveAttribute('aria-label', 'Loading games');
    });

    it('should have default aria-label when not provided', () => {
      render(<SkeletonLoader variant="games" />);
      const skeleton = screen.getByRole('status');
      expect(skeleton).toHaveAttribute('aria-label', 'Loading...');
    });

    it('should include screen reader text', () => {
      render(<SkeletonLoader variant="games" ariaLabel="Loading games" />);
      const srText = screen.getByText('Loading games');
      expect(srText).toBeInTheDocument();
      expect(srText).toHaveClass('sr-only');
    });
  });

  describe('Multiple skeletons', () => {
    it('should render multiple skeletons with spacing', () => {
      const { container } = render(<SkeletonLoader variant="games" count={3} />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('space-y-4'); // Vertical spacing
    });

    it('should render each skeleton with unique key', () => {
      const { container } = render(<SkeletonLoader variant="agents" count={5} />);
      const skeletons = container.querySelectorAll('[role="status"]');
      expect(skeletons).toHaveLength(5);
    });
  });

  describe('Snapshot tests', () => {
    it('should match snapshot for games variant', () => {
      const { container } = render(<SkeletonLoader variant="games" count={2} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for agents variant', () => {
      const { container } = render(<SkeletonLoader variant="agents" count={3} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for message variant', () => {
      const { container } = render(<SkeletonLoader variant="message" count={2} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for chatHistory variant', () => {
      const { container } = render(<SkeletonLoader variant="chatHistory" count={4} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
