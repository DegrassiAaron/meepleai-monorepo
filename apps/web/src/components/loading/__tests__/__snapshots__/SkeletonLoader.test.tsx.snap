// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SkeletonLoader Snapshot tests should match snapshot for agents variant 1`] = `
<div
  class="space-y-4"
>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-20 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-3"
    >
      <div
        class="w-12 h-12 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0"
      />
      <div
        class="flex-1 space-y-2"
      >
        <div
          class="h-4 bg-slate-300 dark:bg-slate-600 rounded w-1/2"
        />
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-20 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-3"
    >
      <div
        class="w-12 h-12 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0"
      />
      <div
        class="flex-1 space-y-2"
      >
        <div
          class="h-4 bg-slate-300 dark:bg-slate-600 rounded w-1/2"
        />
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-20 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-3"
    >
      <div
        class="w-12 h-12 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0"
      />
      <div
        class="flex-1 space-y-2"
      >
        <div
          class="h-4 bg-slate-300 dark:bg-slate-600 rounded w-1/2"
        />
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SkeletonLoader Snapshot tests should match snapshot for chatHistory variant 1`] = `
<div
  class="space-y-4"
>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-12 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-2"
    >
      <div
        class="w-6 h-6 bg-slate-300 dark:bg-slate-600 rounded flex-shrink-0"
      />
      <div
        class="flex-1 space-y-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
        <div
          class="h-2 bg-slate-300 dark:bg-slate-600 rounded w-1/3"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-12 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-2"
    >
      <div
        class="w-6 h-6 bg-slate-300 dark:bg-slate-600 rounded flex-shrink-0"
      />
      <div
        class="flex-1 space-y-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
        <div
          class="h-2 bg-slate-300 dark:bg-slate-600 rounded w-1/3"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-12 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-2"
    >
      <div
        class="w-6 h-6 bg-slate-300 dark:bg-slate-600 rounded flex-shrink-0"
      />
      <div
        class="flex-1 space-y-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
        <div
          class="h-2 bg-slate-300 dark:bg-slate-600 rounded w-1/3"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-12 rounded-md
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-center p-3 space-x-2"
    >
      <div
        class="w-6 h-6 bg-slate-300 dark:bg-slate-600 rounded flex-shrink-0"
      />
      <div
        class="flex-1 space-y-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
        />
        <div
          class="h-2 bg-slate-300 dark:bg-slate-600 rounded w-1/3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SkeletonLoader Snapshot tests should match snapshot for games variant 1`] = `
<div
  class="space-y-4"
>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-64 rounded-lg
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex flex-col p-4 space-y-3"
    >
      <div
        class="flex-1 bg-slate-300 dark:bg-slate-600 rounded"
      />
      <div
        class="h-4 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
      />
      <div
        class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full"
      />
      <div
        class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-5/6"
      />
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-64 rounded-lg
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex flex-col p-4 space-y-3"
    >
      <div
        class="flex-1 bg-slate-300 dark:bg-slate-600 rounded"
      />
      <div
        class="h-4 bg-slate-300 dark:bg-slate-600 rounded w-3/4"
      />
      <div
        class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full"
      />
      <div
        class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-5/6"
      />
    </div>
  </div>
</div>
`;

exports[`SkeletonLoader Snapshot tests should match snapshot for message variant 1`] = `
<div
  class="space-y-4"
>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-16 rounded-xl
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-start p-3 space-x-3"
    >
      <div
        class="w-10 h-10 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0"
      />
      <div
        class="flex-1 space-y-2 pt-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full"
        />
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-4/5"
        />
      </div>
    </div>
  </div>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="
            h-16 rounded-xl
            animate-pulse
            bg-slate-200 dark:bg-slate-700
            
          "
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <div
      class="h-full flex items-start p-3 space-x-3"
    >
      <div
        class="w-10 h-10 bg-slate-300 dark:bg-slate-600 rounded-full flex-shrink-0"
      />
      <div
        class="flex-1 space-y-2 pt-1"
      >
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-full"
        />
        <div
          class="h-3 bg-slate-300 dark:bg-slate-600 rounded w-4/5"
        />
      </div>
    </div>
  </div>
</div>
`;
